### PR TITLE
Updating documentation and guidelines

### PR DESCRIPTION
**Pull request by jenssegers**

### CODE REVIEW

File: apiary.apib [hunk 1]
```diff
@@ -147,6 +134,11 @@ The following values should be passed as POST parameters:
  - `grant_type` - must be "authorization_code"
  - `redirect_uri` - the original submitted redirect URL
 
+Please note that a `content-type` header is **required** for this request. These types are supported:
+
+ - application/x-www-form-urlencoded
```
Comment (Steerlin) [inline apiary.apib]: this conflicts with the above stated : `Data is always passed as JSON` ?

File: apiary.apib [hunk 2]
```diff
@@ -467,27 +459,23 @@ Get a list of contacts.
...
-                    + _items (array)
-                        + (object)
-                            + position: `CEO` (string)
-                            + decision_maker: true (boolean)
-                            + company (object)
-                                + _type: `company` (string)
-                                + id: 123 (string)
+                + tags: tag1, tag2 (array[string])
+                + custom_fields (array)
+                    + (object)
```
Comment (mrtus) [inline apiary.apib]: I like this more then the `_items` wrapped one! 👍

File: apiary.apib [hunk 3]
```diff
@@ -87,7 +73,7 @@ Errors are always returned as an array of error objects under an `errors` top le
 {
     "errors": [
         {
-            "status": 400,
+            "code": 9000,
```
[the remark this answers is not in the document]
Comment (stivni) [reply apiary.apib]: status codes are part of the http protocol, not of the data...

File: apiary.apib [hunk 4]
```diff
@@ -15,15 +15,15 @@ Before creating a new integration, be sure to check out our **[Marketplace](http
 
 ## General principles
 
-### RPC
+### Endpoints
 
 Our API consists of HTTP RPC-style methods, in the form of `https://api.teamleader.eu/resource.action`.
```
Comment (stivni) [inline apiary.apib]: `Our API consists of HTTP endpoints, reflecting actions on a resource.` ?

File: apiary.apib [hunk 5]
```diff
@@ -15,15 +15,15 @@ Before creating a new integration, be sure to check out our **[Marketplace](http
...
 
 We chose this RPC-style approach over the more popular REST, because it enables us to have domain related actions on resources such as `invoices.book`, `timetracking.start` and `timetracking.stop`.
```
Comment (stivni) [inline apiary.apib]: `action-focussed approach`

File: apiary.apib [hunk 6]
```diff
@@ -96,7 +82,7 @@ Errors are always returned as an array of error objects under an `errors` top le
 
 ### Status codes
 
-When objects are created, we return a `201` response containing the id of the resource. When objects are updated or actions are performated, we othen return an empty response with a `204` status code.
+When objects are created, we return a `201` response containing the id of the resource. When objects are updated or actions are performated, we often return an empty response with a `204` status code.
```
Comment (stivni) [inline apiary.apib]: `performated` ?

File: docs/guidelines.md [hunk 7]
```diff
@@ -67,118 +65,94 @@ POST deals.list
...
-POST contacts.tag
-{
-  "id": "e8e4510d-87f7-49be-8d88-d13dd03e23b4",
-  "tags": ["foo", "bar", "baz"]
-}
-```
-
-**note** : currently we don't want this for ids, as we don't know how to cope with transactionality yet.
+ - Not have unexpected side effects
+ - Respons with an empty response and the `204` HTTP status code
```
Comment (stivni) [inline docs/guidelines.md]: Respons => respond

File: docs/guidelines.md [hunk 8]
```diff
@@ -67,118 +65,94 @@ POST deals.list
...
 ```json
 {
-  "id": 123,
+  "id": "ee72ae60-d4df-4f27-beec-d105d5b3e170",
   "name": "John Doe",
   "email": null
 }
 ```
 
 ### Ids
```
Comment (stivni) [inline docs/guidelines.md]: This is now twice in the docs

File: apiary.apib (hunk 3 above)
[the remark this answers is not in the document]
Comment (mrtus) [reply apiary.apib]: 🆗

File: docs/guidelines.md (hunk 8 above)
Comment (jenssegers) [reply docs/guidelines.md]: :ok:

File: docs/guidelines.md [hunk 9]
```diff
@@ -115,15 +116,21 @@ Endpoints used to update existing objects **MUST**:
 
 ## Designing properties
 
+### Adding and modifying endpoints
+
+Adding endpoints and properties are typically always non-breaking and backwards compatible. Modifying, renaming and removing endpoints and properties is allowed during the beta period is allowed. Once our API is released, we will tag our API with a new version for each breaking change.
```
Comment (mrtus) [inline docs/guidelines.md]: You created a weird sentence here `properties is allowed during the beta period is allowed.`

File: docs/guidelines.md (hunk 7 above)
Comment (jenssegers) [reply docs/guidelines.md]: :ok:

File: apiary.apib (hunk 6 above)
Comment (jenssegers) [reply apiary.apib]: :ok:

File: apiary.apib [hunk 10]
```diff
@@ -96,7 +82,7 @@ Errors are always returned as an array of error objects under an `errors` top le
 
 ### Status codes
 
-When objects are created, we return a `201` response containing the type and the id of the resource. When objects are updated or actions are performated, we othen return an empty response with a `204` status code.
+When objects are created, we return a `201` response containing the id of the resource. When objects are updated or actions are performed, we often return an empty response with a `204` status code.
```
Comment (stivni) [inline apiary.apib]: we might still set `type` too, right?

Comment (jenssegers) [reply apiary.apib]: 🆗

File: docs/guidelines.md [hunk 11]
```diff
@@ -67,118 +65,97 @@ POST deals.list
...
+ - Return pagination information in the `meta` response attribute
 
-### Editing, renaming or deleting endpoints
+### Creating objects
 
-- as long as we don't have a versioning system, this is **NOT** allowed
+Endpoints used to create new objects **MUST**:
 
-### Queries
+ - Return the new uuid as a top-level `id` attribute, and the `type` of the object
```
[the remark this answers is not in the document]
Comment (jenssegers) [reply docs/guidelines.md]: :ok:

File: docs/guidelines.md (hunk 9 above)
Comment (stivni) [reply docs/guidelines.md]: _Modifying, renaming and removing endpoints and properties_ is allowed ...

 :ok: by @stivni on 07/11/2017 17:41:23

**Pull request merged**